storage: backpressure SSTs before other writes if L0 fills up

dt opened the pull request:
Currently we start compacting as soon as there are two or more files in
L0. If we reach 20, RocksDB will start slowing all writes in an attempt
to help the compactor catch up.

This global slowdown will affect foreground traffic, liveness, etc and
could significantly impact latency of live traffic or even availability.

Directly ingesting large numbers of small files could make it easier to
hit this limit than it usually is with normal writes -- normal writes
are buffered in the memtable and flushed all at once but direct
ingestion could add potentially lots of files rather quickly.

Additionally, SST ingetions are currently only done by bulk operations
like RESTORE, IMPORT or index backfills. These are all less latency
sensitive then foreground traffic, so we'd prefer to backpressure them
before we risk approaching the global slowdown trigger.

Release note (performance improvement): backpressure bulk operations before other traffic.

dt committed:
storage: backpressure SSTs before other writes if L0 fills up

Currently we start compacting as soon as there are two or more files in
L0. If we reach 20, RocksDB will start slowing all writes in an attempt
to help the compactor catch up.

This global slowdown will affect foreground traffic, liveness, etc and
could significantly impact latency of live traffic or even availability.

Directly ingesting large numbers of small files could make it easier to
hit this limit than it usually is with normal writes -- normal writes
are buffered in the memtable and flushed all at once but direct
ingestion could add potentially lots of files rather quickly.

Additionally, SST ingetions are currently only done by bulk operations
like RESTORE, IMPORT or index backfills. These are all less latency
sensitive then foreground traffic, so we'd prefer to backpressure them
before we risk approaching the global slowdown trigger.

Release note (performance improvement): backpressure bulk operations before other traffic.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34258)
<!-- Reviewable:end -->

dt commented:
This was free-friday foray a bit outside my familiar territory, so hoping to get some guidance from folks who know this code a bit better if this is the right place/approach, in particular w.r.t. what we need to be worried about during apply, blocking raft, etc.

Also, I think ideally we'd only back pressure SSTs that are headed for L0? Pushing this down to Rocks could make sense --after all it is its very similar backpressure mechanism that motivates this, it is already going to have to figure out where this SST is going and it knows how many SSTs it has so it certainly might fit better there. That said, for now I figured I'd like to log when it happens (and maybe even maintain a metric) so doing it separately seemed to make sense?

petermattis reviewed (commented):
This is interesting. Have you done any experiments to see if it makes a difference?

Do we have metrics on how often sstable ingestion is going into L0 vs lower levels? I'm pretty sure no, but I'd like to know those numbers in order to motivate this work. Adding a few fprintfs to RocksDB and running an ingestion-heavy workload (restore or import) would be one way to gather this. 

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-:-LXAglr89v5iEusEUXK5:bdf75ny)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @petermattis, and @tbg)



<!-- Sent from Reviewable.io -->

dt commented:
I did some tests last month in a three-node cluster and did not trigger the RocksDB slowdown -- I had `AddFile(L0 Files)` showing a few thousand, but still had `0 level0_slowdown`. I don't really _expect_ to hit this, but my I think a larger clusters, worst-case data distribution and poor timings could cause a sudden spike that would be pretty bad, so I more wanted something like this as a backstop -- we don't want to hit it, but if we do, hopefully it can take the sting out of an unusual spike.

We don't _currently_ have a problem with this at all: RESTORE and IMPORT should usually ingest only at the lowest level -- since we carve out new key space that isn't seeing any other traffic, the AddSSTAble is usually the first write and overlaps with nothing.

With index backfills though this will not be the case -- a table with heavy write traffic will likely have index entries written by live traffic in the ranges by the time the backfill gets to them, so the SST would overlap pretty recent (i.e. not yet compacted) writes. But that isn't the really scary part: every table reader will be generate batches of index entries, chunking them into SSTs and then shipping those off to the appropriate range. Thus a given range should expect, in the worst case, to receive an SSTs from every reader from every batch (i.e. 100k rows of the source table) -- all of which can overlap each other, forcing them all into L0.

Since a range should expect, worst case, an SST from every producer, a given range should expect to be hit with num-nodes overlapping SSTs at the same time. In a larger cluster, this could hit the constant limit of 20 that triggers the slowdown by itself. The risk seems exacerbated by the fact the node likely has many ranges in the same situation in that RocksDB, so it could easily spike above even the hard stop limit.

tbg reviewed (commented):
> Adding a few fprintfs to RocksDB and running an ingestion-heavy workload (restore or import) would be one way to gather this.

Before we do a one-off here, it wouldn't be more work to add sensible metrics that also make sense in other contexts. For example, we could add metrics `lsm_num_files_l{0,1,2,3,4,5,6}`.

I generally think this change is a good idea, but like @petermattis I think we should find a scenario in which it fires (perhaps the 32 node restore tests?).

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-:-LXJC37dBM2DRLRY8BpW:b-nhltaa)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @nvanbenschoten, @petermattis, and @tbg)

___
*[pkg/storage/engine/rocksdb.go, line 2930 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-LXJBpxL3473ukVfuMGD:-LXJBpxMD4mqq97zXd4s:b-y10yzf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1aeac0b66ca8ef877859c8724eeb6af74765a6b1/pkg/storage/engine/rocksdb.go#L2930)):*
> ```Go
> 	// SST additions well before then at 10.
> 	const l0SlowdownFiles = 10
> 	for i, re := 1, retry.StartWithCtx(ctx, retry.Options{MaxRetries: 6}); re.Next(); i++ {
> ```

Could you specify the options a bit more? Now this is using some defaults but those could change (plus I don't know what to expect here)


<!-- Sent from Reviewable.io -->

danhhz commented:
> RESTORE and IMPORT should usually ingest only at the lowest level -- since we carve out new key space that isn't seeing any other traffic, the AddSSTAble is usually the first write and overlaps with nothing.

Because we keep all replicas in the same rocksdb, this stops being true if, after you kick off the restore/import, you make a new table and give it write traffic.

dt commented:
A customers trying out the latest 19.1 beta hit a rocksdb stall crash and looking at the log they pasted, that node had 326 SSTs added to L0 in the 100s interval prior to the crash, with 536 l0_slowdown stalls and 17k pending compaction byte stalls.

It seems like we might want to dust off this change -- we're working on bigger fixes for 19.2 that might make fewer files, or make us able to ingest more files and ingest faster -- but this change should at least protect the running node from being crashed by excessive AddSSTables.

tbg commented:
@ajkr this topic is probably going to keep us busy in the next release cycle (these stalls have caused node liveness failure and with that, complete temporary cluster failure of course on top of a myriad of other symptoms that tend to really confuse the untrained eye). What's your take on changes like these? Is this a good workaround for now? Is there something else that's not too invasive that we should be looking at?

dt commented:
I think we need to extend this change to also check `pending_compaction_bytes` since I think we can easily hit that too if a sudden spike in ingestions makes any one level, not just L0, disproportionately full.

dt commented:
I added `pendingCompactionBytesEstimate` check to this... does 16gb / 25% of the slowdown threshold seem like a reasonable point to start slowing sst ingestion?

ajkr commented:
That sounds reasonable to me.

@tbg I think backpressure is good since it prevents file ingestion from impacting live traffic. I know @dt has some ideas regarding cutting the files before they're ingested in a way that will reduce overlap, allowing us to ingest them directly into a lower level in some cases. But until that's available, backpressure sounds like a good strategy to me.

dt commented:
I think we should strongly consider backporting this to 19.1.x -- I don't have a reproduction of the stalls the customer saw yet but as it stands, this is by my the best idea I have to preventing worse-case ingestions from turning into stability issues. The worse this change can do, as I understand it it, is slow down SST applications which will hold the replica mu and is thus bad... but probably less bad than holding the _whole store_ behind a rocks stall after the ingestion?

I'll keep trying to reproduce the stall the customer saw to see if I can confirm this works, but I don't have any better ideas right now.

dt commented:
Rebased this and moved the proposer delay to also be outside Eval, so we don't need to hold latches either. The application delay still needs to block raft, but there isn't much we can do about that one short of somehow gossiping follower health info up to the leader.

dt commented:
On #36424, @tbg expressed concern about large numbers of files in L0 and @ajkr expressed interest in pursuing back-pressuring file ingestion, so I've revisited this:
- pulling all the SSTs back just to count L0 was dumb, so I added the `num-files-in-level0` property to the existing `GetStats` returned (which I was calling anyway for `PendingCompcationBytes`).
- switched to delaying a variable amount of time based on how many files over the limit we are (100ms/file) up to the passed max.
- passed a much larger max when delaying prior to eval (e.g. admission control) than when applying in raft. We do in theory need back pressure even when applying in raft since a follower of many well rate-limited leaders could still exceed limits applying the sum of their ingestions, but there are downsizes to blocking too long there, so we probably don't want to be as aggressive as we are in the Store.Send back pressure call.

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34258)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @dt, and @nvanbenschoten)

___
*[pkg/storage/replica_proposal.go, line 394 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lbxo1ba13GxSxXYA_wB:-Lbxo1bbAlunKGG5R0_e:b-84t0c1) ([raw file](https://github.com/cockroachdb/cockroach/blob/099724266aa0e3b993f64e3a24b69b34745fc45c/pkg/storage/replica_proposal.go#L394)):*
> ```Go
> 
> 	if r, ok := eng.(*engine.RocksDB); ok {
> 		r.PreIngestDelay(ctx, PreIngestDefaultDelayLimit)
> ```

I think this needs to be `engine.PreIngestDefaultDelayLimit`


<!-- Sent from Reviewable.io -->

dt commented:
@petermattis @ajkr we're pretty confident we need something like this since we're seeing rocks stalls even with the L0 file limit bumped up to 200.

ajkr commented on pkg/storage/engine/rocksdb.go:
```diff
@@ -80,6 +80,30 @@ var rocksdbConcurrency = envutil.EnvOrDefaultInt(
+var ingestDelayPerFile = settings.RegisterDurationSetting(
+	"rocksdb.ingest_backpressure.delay_l0_file",
+	"delay to add to ingestions per file in L0 over the configured limit",
+	time.Millisecond*200,
```
Have you considered deriving these values from RocksDB's rates/limits? For example the Rocks slowdown rate is currently configured at 16MB/s, so if we had a 64MB file to ingest, we'd calculate a four second delay.

ajkr reviewed (commented):
lgtm, just had a question about adding new settings vs. deriving values from existing ones.

bdarnell reviewed (approved):
> Reviewed 5 of 12 files at r3, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34258)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @nvanbenschoten)

___
*[pkg/storage/engine/rocksdb.go, line 104 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lc37lYKBtT5fyB4k14p:-Lc37lYL615BVjYHBMKN:b-bimv74) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ad344243e90dc10fc2b7902ecefd132255f5767/pkg/storage/engine/rocksdb.go#L104)):*
> ```Go
> 	"rocksdb.ingest_backpressure.max_delay",
> 	"maximum amount of time to backpressure ingestions",
> 	time.Second*5,
> ```

Why is this one configurable but the 15s in Store.Send is not? I think I'd prefer to hard-code both and have fewer knobs to turn. 


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34258)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @dt, @lucy-zhang, and @nvanbenschoten)

___
*[pkg/storage/replica_proposal.go, line 394 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lbxo1ba13GxSxXYA_wB:-Lc3FZWm8xPOfqZxZZ8I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/099724266aa0e3b993f64e3a24b69b34745fc45c/pkg/storage/replica_proposal.go#L394)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I think this needs to be `engine.PreIngestDefaultDelayLimit`
</blockquote></details>

Done.
___
*[pkg/storage/engine/rocksdb.go, line 104 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lc37lYKBtT5fyB4k14p:-Lc3Fcyq7_asB6-CebhQ:b-8632ka) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ad344243e90dc10fc2b7902ecefd132255f5767/pkg/storage/engine/rocksdb.go#L104)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why is this one configurable but the 15s in Store.Send is not? I think I'd prefer to hard-code both and have fewer knobs to turn. 
</blockquote></details>

I made it configurable as a cop-out since I wasn't sure of the best value to pick for a constant. I'm conflicted on what to pick here since it blocks raft if we hit this one, but if we don't, rocks later will when it gets swapped.

I could make both settings, but I also felt like I might be making too many knobs.

I guess making one a setting and the other not is similarly strange though, so maybe the Store.Send call should just pass a multiplier to apply to the setting-read timeout?


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34258)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @bdarnell, @lucy-zhang, @nvanbenschoten, and @tbg)

___
*[pkg/storage/engine/rocksdb.go, line 2930 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-LXJBpxL3473ukVfuMGD:-Lc3GEM48IxVzm_iFV4H:b-8oifzw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1aeac0b66ca8ef877859c8724eeb6af74765a6b1/pkg/storage/engine/rocksdb.go#L2930)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Could you specify the options a bit more? Now this is using some defaults but those could change (plus I don't know what to expect here)
</blockquote></details>

The retry loop went away.
___
*[pkg/storage/engine/rocksdb.go, line 92 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-LX5SSfzypb2YzKfFxm9-r4-92:-Lc3GZ3VD_BL_KjNLo58:b-83zdj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ad344243e90dc10fc2b7902ecefd132255f5767/pkg/storage/engine/rocksdb.go#L92)):*
<details><summary><i>Previously, ajkr (Andrew Kryczka) wrote…</i></summary><blockquote>

Have you considered deriving these values from RocksDB's rates/limits? For example the Rocks slowdown rate is currently configured at 16MB/s, so if we had a 64MB file to ingest, we'd calculate a four second delay.
</blockquote></details>

That sounds like it could work though I think currently I'm going the opposite direction and making these runtime-tunable knobs -- (which, as it stands, most of our rocks options are not since they're not hooked up to any knobs a user can get to).


<!-- Sent from Reviewable.io -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=34258) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=34258) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cockroachdb/cockroach?pullRequest=34258) it.</sub>

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34258)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @bdarnell, @lucy-zhang, @nvanbenschoten, and @tbg)

___
*[pkg/storage/engine/rocksdb.go, line 104 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lc37lYKBtT5fyB4k14p:-Lc3zlGYBnALOelNwKI4:bigy48s) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ad344243e90dc10fc2b7902ecefd132255f5767/pkg/storage/engine/rocksdb.go#L104)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I made it configurable as a cop-out since I wasn't sure of the best value to pick for a constant. I'm conflicted on what to pick here since it blocks raft if we hit this one, but if we don't, rocks later will when it gets swapped.

I could make both settings, but I also felt like I might be making too many knobs.

I guess making one a setting and the other not is similarly strange though, so maybe the Store.Send call should just pass a multiplier to apply to the setting-read timeout?
</blockquote></details>

I make them both just call `PreIngestDelay(ctx)` and they both just use the setting value.
I'm hesitant to switch to a constant that hasn't had weeks or month of time on master to bake, because if we get it wrong, I'd rather have the escape hatch of telling people to change the setting as opposed to deploy a new binary.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
The overall approach looks good.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-:-Lc6NvZL8PxWrU_n2Flc:bxohitw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @bdarnell, @dt, @lucy-zhang, @nvanbenschoten, and @tbg)

___
*[c-deps/libroach/engine.cc, line 208 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lc6L6TA6EIYWoAW0uBK:-Lc6L6TA6EIYWoAW0uBL:b-qxg43i) ([raw file](https://github.com/cockroachdb/cockroach/blob/b88b6517333f9dcbb137f976073161ad9bd53760/c-deps/libroach/engine.cc#L208)):*
> ```cpp
>   std::string l0_file_count_str;
>   rep->GetProperty("rocksdb.num-files-at-level0",
>                       &l0_file_count_str);
> ```

Nit: the indentation of the second line seems off. It looks reasonable for this to be on a single line.
___
*[pkg/storage/replica_proposal.go, line 393 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lc6LNL0A8KRECzY69al:-Lc6LNL0A8KRECzY69am:b-5q2jrl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b88b6517333f9dcbb137f976073161ad9bd53760/pkg/storage/replica_proposal.go#L393)):*
> ```Go
> 	}
> 
> 	if r, ok := eng.(*engine.RocksDB); ok {
> ```

Is there a reason you didn't add `PreIngestDelay` is a method on the `Engine` interface? All of these type-assertions to `*engine.RocksDB` will have to go away if we support another storage engine. (I realize there are similar type-assertions already in the code, I'm just reluctant to see more added).
___
*[pkg/storage/engine/rocksdb.go, line 2930 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-LXJBpxL3473ukVfuMGD:-Lc6MRnI6KF2qcg4cs01:bx7w41z) ([raw file](https://github.com/cockroachdb/cockroach/blob/1aeac0b66ca8ef877859c8724eeb6af74765a6b1/pkg/storage/engine/rocksdb.go#L2930)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

The retry loop went away.
</blockquote></details>

Heh, I was expecting some sort of loop here waiting for the number of L0 files to fall below a target count. I'm curious why you moved away from that approach. If something is horking up L0 compactions, I think we'd want to continue to wait until the number of L0 files fell below a threshold (up to some maximum wait time).
___
*[pkg/storage/engine/rocksdb.go, line 86 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lc6N3Ul38W-Ix6zgpJX:-Lc6N3Ul38W-Ix6zgpJY:b-nwl8kk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b88b6517333f9dcbb137f976073161ad9bd53760/pkg/storage/engine/rocksdb.go#L86)):*
> ```Go
> 	"rocksdb.ingest_backpressure.l0_file_count_threshold",
> 	"number of L0 files after which to backpressure SST ingestions",
> 	20,
> ```

That's a lot of L0 files. Is there a rationale behind setting this to 20 vs something lower? 
___
*[pkg/storage/engine/rocksdb.go, line 103 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lc6NBybEklXAPiA-zuW:-Lc6NBybEklXAPiA-zuX:bxiif9n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b88b6517333f9dcbb137f976073161ad9bd53760/pkg/storage/engine/rocksdb.go#L103)):*
> ```Go
> var ingestDelayTime = settings.RegisterDurationSetting(
> 	"rocksdb.ingest_backpressure.max_delay",
> 	"maximum amount of time to backpressure ingestions",
> ```

`to backpressure a single ingestion`
___
*[pkg/storage/engine/rocksdb.go, line 3025 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lc6M_tm2qG5DfGv4gMU:-Lc6M_tm2qG5DfGv4gMV:b9los7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/b88b6517333f9dcbb137f976073161ad9bd53760/pkg/storage/engine/rocksdb.go#L3025)):*
> ```Go
> }
> 
> func (r *RocksDB) calculatePreIngestDelay(stats *Stats) time.Duration {
> ```

The only member of `RocksDB` you're using is `RocksDB.cfg`. I think it would be better to pass that as a parameter and make this a global function. That would remove the awkwardness in `TestIngestDelayLimit` of creating a `RocksDB` and only populating a single field.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34258)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @bdarnell, @lucy-zhang, @nvanbenschoten, @petermattis, and @tbg)

___
*[c-deps/libroach/engine.cc, line 208 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lc6L6TA6EIYWoAW0uBK:-Lc6a-HJ7QaT15dTF-gL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b88b6517333f9dcbb137f976073161ad9bd53760/c-deps/libroach/engine.cc#L208)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Nit: the indentation of the second line seems off. It looks reasonable for this to be on a single line.
</blockquote></details>

Done.
___
*[pkg/storage/replica_proposal.go, line 393 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lc6LNL0A8KRECzY69al:-Lc6YtU7BgptJCH1VrLD:bfkdkm4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b88b6517333f9dcbb137f976073161ad9bd53760/pkg/storage/replica_proposal.go#L393)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is there a reason you didn't add `PreIngestDelay` is a method on the `Engine` interface? All of these type-assertions to `*engine.RocksDB` will have to go away if we support another storage engine. (I realize there are similar type-assertions already in the code, I'm just reluctant to see more added).
</blockquote></details>

It seemed like it was working around a RocksDB-specific problem, so the type-specific call-sites didn't bother me, but I could also see other engines just nooping.

Done.
___
*[pkg/storage/engine/rocksdb.go, line 2930 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-LXJBpxL3473ukVfuMGD:-Lc6a6F5D-bYUwdg_lUc:b-pi6ph0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1aeac0b66ca8ef877859c8724eeb6af74765a6b1/pkg/storage/engine/rocksdb.go#L2930)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Heh, I was expecting some sort of loop here waiting for the number of L0 files to fall below a target count. I'm curious why you moved away from that approach. If something is horking up L0 compactions, I think we'd want to continue to wait until the number of L0 files fell below a threshold (up to some maximum wait time).
</blockquote></details>

Yeah, I had a loop like that, but it made the code a tad more complicated and made the service order / queue discipline a bit murky. 
___
*[pkg/storage/engine/rocksdb.go, line 86 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lc6N3Ul38W-Ix6zgpJX:-Lc6U3O4DHHtKj85Tedo:b-8u5bqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b88b6517333f9dcbb137f976073161ad9bd53760/pkg/storage/engine/rocksdb.go#L86)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

That's a lot of L0 files. Is there a rationale behind setting this to 20 vs something lower? 
</blockquote></details>

20 was the old slowdown limit (before we bumped it to 200) which is why I picked it.

While that would be a lot of L0 files if they were normal L0 files -- i.e. the size of full memtable and covering the whole key span and thus adding a full read amplification multiple -- when we're ingesting, we have reason to believe that many of these files are *not* normal L0 files. Ingested files probably only cover a small subspan of the full key span, so the majority of traffic can likely skip then via the bounds check and never end up seeking though the file itself. 

Of course some of the files will be ingest-triggered memtable flushes, which likely still cover the whole active keyspan, so but even those are probably smaller than usual since we're forcing frequent flushes during ingest.

If we back-pressused at a much smaller number, we'd likely bring the ingestion to a crawl if it hits some data distribution that results in many small SSTs. Indeed, generally when we see larger numbers of L0 files, they're usually quite small (<1mb) and we actually kind of want the compactor to get as many of these in a single pass as it can (indeed, I'd like to raise the compaction trigger during ingest, but that should happen separately), so letting a few more files into L0 should help the ingest continue to make progress.

All of that said, I don't actually know if 20 is the right number, which is part of why I made it a dynamically adjustable setting.
___
*[pkg/storage/engine/rocksdb.go, line 103 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lc6NBybEklXAPiA-zuW:-Lc6ZW4k8fB6SkGFzlsf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b88b6517333f9dcbb137f976073161ad9bd53760/pkg/storage/engine/rocksdb.go#L103)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`to backpressure a single ingestion`
</blockquote></details>

Done.
___
*[pkg/storage/engine/rocksdb.go, line 3025 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lc6M_tm2qG5DfGv4gMU:-Lc6b7OXAf4WcPIG8qAL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b88b6517333f9dcbb137f976073161ad9bd53760/pkg/storage/engine/rocksdb.go#L3025)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The only member of `RocksDB` you're using is `RocksDB.cfg`. I think it would be better to pass that as a parameter and make this a global function. That would remove the awkwardness in `TestIngestDelayLimit` of creating a `RocksDB` and only populating a single field.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (approved):
> Reviewed 7 of 7 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34258)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @lucy-zhang, @nvanbenschoten, @petermattis, and @tbg)

___
*[pkg/storage/engine/rocksdb.go, line 92 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-LX5SSfzypb2YzKfFxm9-r4-92:-Lc6o1Bu5QY5FWbPxaWi:bcoy3wd) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ad344243e90dc10fc2b7902ecefd132255f5767/pkg/storage/engine/rocksdb.go#L92)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

That sounds like it could work though I think currently I'm going the opposite direction and making these runtime-tunable knobs -- (which, as it stands, most of our rocks options are not since they're not hooked up to any knobs a user can get to).
</blockquote></details>

Note that we do allow arbitrary rocksdb options to be passed as `--store=path=/foo,rocksdb=key1=val1;key2=val2`. 


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34258)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @dt, @lucy-zhang, @nvanbenschoten, and @tbg)

___
*[pkg/storage/engine/rocksdb.go, line 86 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34258#-Lc6N3Ul38W-Ix6zgpJX:-Lc75YGREJznoVgZpTRu:b-2meck3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b88b6517333f9dcbb137f976073161ad9bd53760/pkg/storage/engine/rocksdb.go#L86)):*
Ack. Let's leave as-is for now. 

> so the majority of traffic can likely skip then via the bounds check and never end up seeking though the file itself

Iterators still need to consider each of these files, even though they are relatively quick checks. I'm not sure how that plays out in practice, though.


<!-- Sent from Reviewable.io -->

dt commented:
bors r+

dt commented:
bors r+
looks like bors forgot about this one

RaduBerinde commented:
This PR adds an unused setting `ingestDelayPerFile`. Ok to remove?

RaduBerinde commented:
@dt Any idea about `ingestDelayPerFile`?
